### PR TITLE
Match with max width

### LINKED ISSUE #1838: bad formatting for match pattern list that almost fits on one line
Rustfmt leaves all the different tags on the same line even if this means the `=>` wide arrow does not fit after them. The following snippet gets the error `Rustfmt failed at stdin:3: line exceeded maximum length (maximum: 100, found: 103) (sorry)`

```rust
fn bad_formatting() -> bool {
    match m {
        Variant::Tag | Variant::Tag2 | Variant::Tag3 | Variant::Tag4 | Variant::Tag5 | Variant::Tag6 =>
        {}
    }
}
```

### LINKED ISSUE #1839: off by one error leads to unnecessary extra line
The second `let` statement below should be formatted like the first `let` statement, it still fits inside the maximum allowed length.

```rust
fn bad_formatting() {
    let (this_line_uses_99_characters_and_is_formatted_properly, x012345) = match some_expression {
        _ => unimplemented!(),
    };

    let (should_be_formatted_like_the_line_above_using_100_characters, x0) =
        match some_expression {
            _ => unimplemented!(),
        };
}
```

### PR DESCRIPTION
**Pull request by topecongiro**
Closes #1838.
Closes #1839.

### CODE REVIEW

File: src/expr.rs [hunk 1]
```diff
@@ -494,6 +494,7 @@ where
             definitive_tactic(
                 &items,
                 ListTactic::LimitedHorizontalVertical(context.config.array_width()),
+                2,
```
Comment (nrc) [inline src/expr.rs]: Could you comment on where the `2` comes from please? (Here and above).

File: src/expr.rs [hunk 2]
```diff
@@ -1668,7 +1682,7 @@ fn rewrite_match_pattern(
     );
 
     let items: Vec<_> = pat_strs.into_iter().map(ListItem::from_str).collect();
-    let tactic = definitive_tactic(&items, ListTactic::HorizontalVertical, pat_shape.width);
+    let tactic = definitive_tactic(&items, ListTactic::HorizontalVertical, 3, pat_shape.width);
```
Comment (nrc) [inline src/expr.rs]: and the `3`

**Comment by topecongiro** [comment]
@nrc Thank you for the review! Updated and rebased with using enum.

**Comment by nrc** [comment]
Thanks, I like the change!

**Pull request merged**